### DISCUSSION TITLE
Standalone MinecraftServer usage

### DISCUSSION THREAD

**Discussion by h-wb** [jeremylvln/Shulker#340] [Q&A] [answered]
Hey!
I'm playing around with Shulker, and I just want to let you know that I might do a bunch of out of scope shit so don't hesitate to tell me that i'm not using it the right way ahaha... 
I basically want to use it to run a vanilla server and a modded server behind the proxy (if that's even possible 🤔) but will probably not take advantage of the scaling capabilities. Using a regular Deployment of the itgz server would probably be a better fit but over-engineering is fun..

Anyway, this might be a better fit for a bug report but is it actually possible to run a MinecraftServer without a MinecraftServerFleet ? You can see my relevant config files [here](https://github.com/h-wb/home-ops/tree/fb6cf9bfcb8ab73c980a806f0ebf53ca0a735d6a/kubernetes/apps/games/shulker/cluster)

 When I'm running a MinecraftServerFleet and one standalone MinecraftServer, I can successfully connect to the proxy and can connect to both servers:
 
<img width="627" alt="image" src="https://github.com/jeremylvln/Shulker/assets/19188789/fdb0bdaf-dc77-47c3-95de-35fca6060b03">

Then I tried staying connected to the 'survival' server while removing the MinecraftServerFleet 'lobby': 
<img width="465" alt="image" src="https://github.com/jeremylvln/Shulker/assets/19188789/5f39fc75-9710-4f68-8a46-6aa09f29f494">

I'm still connected (I see the 'lobby' server tag but not the actual minecraft server anymore)

If I disconnect and try to reconnect again tho, I can't : 
<img width="877" alt="image" src="https://github.com/jeremylvln/Shulker/assets/19188789/9b4947b7-6716-45cf-9ec8-b747c764b92c">

I don't know if it' relevant but here's the proxy logs : 

```
io.netty.channel.AbstractChannel$AnnotatedConnectException: finishConnect(..) failed: Connection refused: localhost/127.0.0.1:30000
Caused by: java.net.ConnectException: finishConnect(..) failed: Connection refused
	at io.netty.channel.unix.Errors.newConnectException0(Errors.java:166) ~[velocity-3.3.0-SNAPSHOT-320.jar:3.3.0-SNAPSHOT (git-08c03aae-b320)]
	at io.netty.channel.unix.Errors.handleConnectErrno(Errors.java:131) ~[velocity-3.3.0-SNAPSHOT-320.jar:3.3.0-SNAPSHOT (git-08c03aae-b320)]
	at io.netty.channel.unix.Socket.finishConnect(Socket.java:359) ~[velocity-3.3.0-SNAPSHOT-320.jar:3.3.0-SNAPSHOT (git-08c03aae-b320)]
	at io.netty.channel.epoll.AbstractEpollChannel$AbstractEpollUnsafe.doFinishConnect(AbstractEpollChannel.java:710) ~[velocity-3.3.0-SNAPSHOT-320.jar:3.3.0-SNAPSHOT (git-08c03aae-b320)]
	at io.netty.channel.epoll.AbstractEpollChannel$AbstractEpollUnsafe.finishConnect(AbstractEpollChannel.java:687) ~[velocity-3.3.0-SNAPSHOT-320.jar:3.3.0-SNAPSHOT (git-08c03aae-b320)]
	at io.netty.channel.epoll.AbstractEpollChannel$AbstractEpollUnsafe.epollOutReady(AbstractEpollChannel.java:567) ~[velocity-3.3.0-SNAPSHOT-320.jar:3.3.0-SNAPSHOT (git-08c03aae-b320)]
	at io.netty.channel.epoll.EpollEventLoop.processReady(EpollEventLoop.java:499) ~[velocity-3.3.0-SNAPSHOT-320.jar:3.3.0-SNAPSHOT (git-08c03aae-b320)]
	at io.netty.channel.epoll.EpollEventLoop.run(EpollEventLoop.java:407) ~[velocity-3.3.0-SNAPSHOT-320.jar:3.3.0-SNAPSHOT (git-08c03aae-b320)]
	at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:997) ~[velocity-3.3.0-SNAPSHOT-320.jar:3.3.0-SNAPSHOT (git-08c03aae-b320)]
	at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74) ~[velocity-3.3.0-SNAPSHOT-320.jar:3.3.0-SNAPSHOT (git-08c03aae-b320)]
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30) ~[velocity-3.3.0-SNAPSHOT-320.jar:3.3.0-SNAPSHOT (git-08c03aae-b320)]
	at java.lang.Thread.run(Thread.java:833) [?:?]
```

EDIT:

After a not so scientific experiment I think I found a case where it's working, by following that naming convention:

```
apiVersion: shulkermc.io/v1alpha1
kind: MinecraftServer
metadata:
  name: lobby-whatever
spec:
  clusterRef:
    name: minecraft-cluster
  tags:
    - lobby
```

I do have another question if you don't mind, i've been trying to set-up serverProperties without success, maybe i'm using the wrong syntax :

```
    serverProperties:
      enforce-whitelist: "true"
```

**Comment by jeremylvln** [comment]
Hi @h-wb!

Thanks for using Shulker, it means a lot to me :)

You can clearly only use `MinecraftServer` resources if you want, without any scaling capabilities. You can mix the kind and versions of your servers, as long as the proxy is able to handle them. Please note that Shulker needs a companion add-on to the servers, so you can't use anything but only the platforms Shulker's supports. Currently, only Bukkit-based servers are supported (with a minimum of Paper, so upstream forks like Folia should work too). I don't want to focus my time on creating a mod agent (either for Forge or Fabric), so I think you'll not be able to use a modded server for now.

You need at least one server tagged with `lobby` to be able to connect to your Minecraft cluster. Shulker will automatically send your players to one of the servers with this tag. So if there is none, you'll not be able to connect, and you'll get the screen you gave. The `limbo` mechanism is really not finalized, so don't pay much attention to the message itself.

About the server properties, it's weird. I've automated tests that confirm it's working, but I've never tested myself. Do you confirm you use something like the following?

```yaml
apiVersion: shulkermc.io/v1alpha1
kind: MinecraftServer
metadata:
  name: lobby-whatever
spec:
  clusterRef:
    name: minecraft-cluster
  tags:
    - lobby
  config:
    serverProperties:
      enforce-whitelist: "true"
```

After your server is ready, you can open a terminal inside it (`sh` or `bash`) and you can check the content of your `server.properties`.

**Reply by h-wb** [reply]
> Currently, only Bukkit-based servers are supported (with a minimum of Paper, so upstream forks like Folia should work too). I don't want to focus my time on creating a mod agent (either for Forge or Fabric), so I think you'll not be able to use a modded server for now.

Aah that makes sense, yeah I was not familiar with the different servers implementations. Forge doesn't look like the best to work with ahaha

> You need at least one server tagged with lobby to be able to connect to your Minecraft cluster. Shulker will automatically send your players to one of the servers with this tag. So if there is none, you'll not be able to connect, and you'll get the screen you gave. The limbo mechanism is really not finalized, so don't pay much attention to the message itself.

Indeed, no problem with a MinecraftServer named `lobby-xyz` and it was actually written in the doc sorry about that!

> About the server properties, it's weird. I've automated tests that confirm it's working, but I've never tested myself. Do you confirm you use something like the following?

Yep, I can't seem to make it work, i've been checking the content of the `server.properties` but it's not overridden. Your test suite looks super solid so i'm gonna double check 😅 I can also try to use the ConfigMap (I've also been trying to use the itgz docker image env variables without success as well)

**Reply by jeremylvln** [reply, marked as answer]
There was indeed a bug with the `server.properties`. I'm pushing a fix right now and I will release a new version just after :)

**Reply by jeremylvln** [reply]
The fix is now released :)

**Reply by h-wb** [reply]
It works! Thanks for the quick fix 🙏
